#+CATEGORY: luad

* Objective

To perform a multi-gene epistasis analysis of as big of a LUAD dataset
as we can assemble (WGS+WXS+TGS), evaluating the impact of smoking on
the set of mutations expected to contribute to progression, after
mutation of TP53 and KRAS.

* Requirements

- Python, ideally installed with pyenv
  (https://github.com/pyenv/pyenv) to be able to have exactly the same
  version (3.9.5) as the tested, but otherwise at least 3.7 so that
  the order of dictionaries is preserved.
- R (in debian or ubuntu install the packages r-base and r-base-dev)
- the development files for libcurl (OpenSSL flavour) are required for
  many R packages including RCurl (that is required for
  cancereffectsizeR). In debian or ubuntu install the
  libcurl4-openssl-dev package

* Setup

To prepare the system for running the analysis, within the ==code==
directory run
#+begin_src shell
  ./setup.sh
#+end_src
and follow the instructions.

* Data decisions

- TCGA has smoking as pack years. It has NA values but not 0. We will
  be assuming that NA means no smoking, i.e., that there was
  information for smoking for all participants

* Sources of genes in consideration

- [[https://www.sciencedirect.com/science/article/pii/S1470204515000777#][Genomic alterations in lung adenocarcinoma]]
- [[https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3092285/][The Genomics of Lung Adenocarcinoma]]
- [[https://www.nature.com/articles/nature13385][Comprehensive molecular profiling of lung adenocarcinoma]]
- [[https://genome.cshlp.org/content/22/11/2109.full][The transcriptional landscape and mutational profile of lung adenocarcinoma]]
- [[https://www.sciencedirect.com/science/article/pii/S1556086416309273][Comprehensive Characterization of Oncogenic Drivers in Asian Lung Adenocarcinoma]]

* TODO [0/3] Before publishing                                      :waiting:

- [ ] Check differences between cancer_epistasis theory.py and
  cancer_epistasis.py

- [ ] Remove access to the files at misc.cidma.us/data and provide
  instructions on how to download each file that are is not accessible
  via cBioPortal or synapse.

- [ ] Check with Stephen status of package lift_coords and change it
  from the requirements (currently requires access to a private repo).

* Introduction

"If our aim is to understand cancer evolution, we cannot ignore the
environment in which a cancer cell evolves." Genetic and non-genetic
clonal diversity in cancer evolution by James R. M. Black and Nicholas
McGranahan

* For introduction

There is a need for using information about co-ocurrence of mutations
in a clinical setting, particularly of KRAS occurs with other
mutations in lung cancer (although this is NSCLC): K-ras Mutation
Subtypes in NSCLC and Associated Co-occuring Mutations in Other
Oncogenic Pathways
https://www.jto.org/article/S1556-0864(18)33532-9/fulltext

* For discussion

A paragraph about copy numbers, or just check comparing results with
copy number with or without a lot(?).

This paper didn't find a difference between smokers and non-smokers in
the co-ocurrence of mutationss with KRAS: K-ras Mutation Subtypes in
NSCLC and Associated Co-occuring Mutations in Other Oncogenic Pathways
https://www.jto.org/article/S1556-0864(18)33532-9/fulltext


More on KRAS
https://www.sciencedirect.com/science/article/pii/S0923753423006567

Differences in selection might caused directly by something other than
smoking, for example non-smokers are older and aging could be directly
responsible for most of the difference in selection between smokers
and non-smokers. But it is clear (is it?) that smoking is the factor
that causes those differences.


K-ras Mutation Subtypes in NSCLC and Associated Co-occuring Mutations
in Other Oncogenic Pathways
https://www.jto.org/article/S1556-0864(18)33532-9/fulltext


TP53 mutations are associated with the development of distant
metastases and poor diagnosis and are more common in smokers: Genomic
analysis of early-stage lung cancer reveals a role for TP53 mutations
in distant metastasis
https://www.nature.com/articles/s41598-022-21448-1


If we take a look at KRAS effect on other mutations and OS: "Mutations
in KEAP1 and SMARCA4, but not STK11, were associated with metastatic
disease and poor OS. Functional STK11 loss, however, may contribute to
poor outcomes in KEAP1MUT tumors." Molecular markers of metastatic
disease in KRAS mutant lung adenocarcinoma
https://www.sciencedirect.com/science/article/pii/S0923753423006567

We can run a model with KRAS, KEAP1, STK11 and maybe SMARCA4 check if
the above is true.


Comparative bioinformatic analysis of KRAS, STK11 and KEAP1
(co-)mutations in non-small cell lung cancer with a special focus on
KRAS G12C
https://www.lungcancerjournal.info/article/S0169-5002(23)00899-1/fulltext


KRAS can be bunched together: "Comparing size distributions of tumors of each tumor sup- pressor
genotype revealed consistent effects between KRAS G12C- and KRAS
G12D-driven lung tumors." Oncogenic context shapes the Ô¨Åtness
landscape of tumor suppression
https://www.nature.com/articles/s41467-023-42156-y

Compare selection coefficients with graph that contains KRAS + TSGs
(Figure 2A)

Check genes that interact differently with EGFR and BRAF and do
pairwise epistasis (maybe + TP53) to compare if we get the same results.

* Possible reviewers

Teresa Landi (suggested by Jeff on the slack channel [2022-08-22 Mon])
